clippy_lints/src/unused_self.rs

```diff
@@ -0,0 +1,102 @@
+use if_chain::if_chain;
+use rustc::hir::def::Res;
+use rustc::hir::intravisit::{walk_path, NestedVisitorMap, Visitor};
+use rustc::hir::{AssocItemKind, HirId, ImplItemKind, ImplItemRef, Item, ItemKind, Path};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_lint_pass, declare_tool_lint};
+
+use crate::utils::span_help_and_lint;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks methods that contain a `self` argument but don't use it
+    ///
+    /// **Why is this bad?** It may be clearer to define the method as a static function instead
+    /// of an instance method if it doesn't require `self`.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust,ignore
+    /// struct A;
+    /// impl A {
+    ///     fn method(&self) {}
+    /// }
+    /// ```
+    ///
+    /// Could be written:
+    ///
+    /// ```rust,ignore
+    /// struct A;
+    /// impl A {
+    ///     fn method() {}
+    /// }
+    /// ```
+    pub UNUSED_SELF,
+    pedantic,
+    "methods that contain a `self` argument but don't use it"
+}
+
+declare_lint_pass!(UnusedSelf => [UNUSED_SELF]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedSelf {
+    fn check_item(&mut self, cx: &LateContext<'a, 'tcx>, item: &Item) {
+        if item.span.from_expansion() {
+            return;
+        }
+        if let ItemKind::Impl(_, _, _, _, None, _, ref impl_item_refs) = item.kind {
+            for impl_item_ref in impl_item_refs {
+                if_chain! {
+                    if let ImplItemRef {
+                        kind: AssocItemKind::Method { has_self: true },
+                        ..
+                    } = impl_item_ref;
+                    let impl_item = cx.tcx.hir().impl_item(impl_item_ref.id);
+                    if let ImplItemKind::Method(_, body_id) = &impl_item.kind;
+                    then {
+                        let body = cx.tcx.hir().body(*body_id);
+                        let self_param = &body.params[0];
+                        let self_hir_id = self_param.pat.hir_id;
+                        let mut visitor = UnusedSelfVisitor {
+                            cx,
+                            uses_self: false,
+                            self_hir_id: &self_hir_id,
+                        };
+                        visitor.visit_body(body);
+                        if !visitor.uses_self {
+                            span_help_and_lint(
+                                cx,
+                                UNUSED_SELF,
+                                self_param.span,
+                                "unused `self` argument",
+                                "consider refactoring to a static method or function",
```

Comment:
"static method" is not a terminology used in rustc or docs at least; we call functions in impl blocks "associated" and then a function without a receiver is a non-method.

Reply:
"static method" is written many times here: https://doc.rust-lang.org/rust-by-example/fn/methods.html

Reply:
Well we used to say "static method" but we do not anymore so it's not surprising that it exists in some old docs but neither the reference nor the book uses it.

Reply:
Yes, now I see, that the `associated function` [is used](https://doc.rust-lang.org/beta/book/ch05-03-method-syntax.html#associated-functions) instead in the official book these days.

So now we have associated constants and associated methods which sounds quite more logical now.